nnfabrik/templates/checkpoint.py

```diff
@@ -0,0 +1,166 @@
+import copy
+from typing import Dict, Tuple
+
+import datajoint as dj
+import tempfile
+import torch
+import os
+from nnfabrik.main import Model, Dataset, Trainer, Seed, Fabrikant
+from nnfabrik.builder import get_all_parts, get_model, get_trainer
+from nnfabrik.templates.trained_model import TrainedModelBase
+from nnfabrik.utility.dj_helpers import make_hash, clone_conn, CustomSchema
+from nnfabrik.builder import resolve_data
+from datajoint.fetch import DataJointError
+from nnfabrik.main import schema
+
+
+def my_checkpoint(nnfabrik):
+    conn_clone = clone_conn(dj.conn())
+    schema_clone = CustomSchema(nnfabrik.schema.database, connection=conn_clone)
+
+    @schema_clone
+    class Checkpoint(dj.Manual):
+        storage = "minio"
+
+        @property
+        def definition(self):
+            definition = """
+            # Checkpoint table
+            -> nnfabrik.Trainer
+            -> nnfabrik.Dataset
+            -> nnfabrik.Model
+            -> nnfabrik.Seed
+            epoch:                             int          # epoch of creation
+            ---
+            score:                             float        # current score at epoch
+            state:                             attach@{storage}  # current state
+            ->[nullable] nnfabrik.Fabrikant
+            trainedmodel_ts=CURRENT_TIMESTAMP: timestamp    # UTZ timestamp at time of insertion
+            """.format(
+                storage=self.storage
+            )
+            return definition
+
+    return Checkpoint
+
+
+class TrainedModelChkptBase(TrainedModelBase):
+    checkpoint_table = None  # Checkpoint
+    keys = [
+        "model_fn",
+        "model_hash",
+        "dataset_fn",
+        "dataset_hash",
+        "trainer_fn",
+        "trainer_hash",
+    ]
+
+    def call_back(self, uid=None, epoch=None, model=None, state=None):
+        """
+        This method is periodically called by the trainer and is used to save the training state in a remote table.
+        Args:
+            uid - Unique identifier for the trained model entry
+            epoch - the iteration count
+            model - current model under training
+            state - Additional information provided by the trainer
+                score: float = 0.0,
+                maximize_score: bool = True,
+                keep_last_n: int = 1,
+                keep_best_n: int = 1,
+                keep_selection: Tuple = (),
+        """
+        maximize_score = state.pop("maximize_score", True)
+        if epoch >= 0:  # save current epoch
```

Comment:
Also I think we wanted to change how the behavior of the callback is determined. So instead of `-2`, `-1` and `0` you could use `"best"`, `"last"` and `"save"`, respectively. This would make it more readable. The string argument could be stored in the `state` argument passed to the callback.